addons/binding/org.openhab.binding.mihome/src/main/java/org/openhab/binding/mihome/internal/socket/XiaomiSocket.java

```diff
@@ -0,0 +1,203 @@
+    private static final JsonParser PARSER = new JsonParser();
+
+    Logger logger = LoggerFactory.getLogger(this.getClass().getName());
+
+    static ArrayList<XiaomiSocket> openSockets = new ArrayList<XiaomiSocket>();
+
+    int port = 0;
+    DatagramSocket socket;
+
+    /**
+     * Sets up an {@link XiaomiSocket} with the MiHome multicast address and a random port
+     *
+     */
+    public XiaomiSocket() {
+        this(0);
+    }
+
+    /**
+     * Sets up an {@link XiaomiSocket} with the MiHome multicast address and a specific port
+     *
+     * @param port - the socket will be bound to this port
+     */
+    public XiaomiSocket(int port) {
+        this.port = port;
+
+        setupSocket();
+    }
+
+    abstract void setupSocket();
+
+    /**
+     * Interrupts the {@link ReceiverThread} and closes the {@link XiaomiSocket}.
+     */
+    private void closeSocket() {
+        synchronized (XiaomiSocket.class) {
+            if (socketReceiveThread != null) {
+                socketReceiveThread.interrupt();
+            }
+            if (socket != null) {
+                socket.close();
+                logger.debug("Socket closed");
+                socket = null;
+            }
+        }
+    }
+
+    /**
+     * Registers a {@link XiaomiSocketListener} to be called back, when data is received.
+     * If no {@link XiaomiSocket} exists, when the method is called, it is being set up.
+     *
+     * @param listener - {@link XiaomiSocketListener} to be called back
+     */
+    public synchronized void registerListener(XiaomiSocketListener listener) {
+        if (socket == null) {
+            setupSocket();
+        }
+        if (!listeners.contains(listener)) {
+            logger.trace("Adding socket listener {}", listener.toString());
+            listeners.add(listener);
+        }
+    }
+
+    /**
+     * Unregisters a {@link XiaomiSocketListener}. If there are no listeners left,
+     * the {@link XiaomiSocket} is being closed.
+     *
+     * @param listener - {@link XiaomiSocketListener} to be unregistered
+     */
+    public synchronized void unregisterListener(XiaomiSocketListener listener) {
+        listeners.remove(listener);
+
+        if (listeners.isEmpty()) {
+            closeSocket();
+        }
+    }
+
+    /**
+     * Sends a message through the {@link XiaomiSocket} to a specific address and port
+     *
+     * @param message - Message to be sent
+     * @param address - Address, to which the message shall be sent
+     * @param port - - Port, through which the message shall be sent
+     */
+    public void sendMessage(String message, InetAddress address, int port) {
+        try {
+            byte[] sendData = message.getBytes("UTF-8");
+            DatagramPacket sendPacket = new DatagramPacket(sendData, sendData.length, address, port);
+            logger.trace("Sending message: {} to {}", message, address.toString());
```

Comment:
You don't have to call toString it will be called automatically, but only when the logging-level actually applies.